### ISSUE TITLE
Unexpected error in LSP server - Source: Ruff (Extension)

### ISSUE THREAD

**Issue by lasinludwig** [charliermarsh/ruff-vscode#58]
Hi there,

I always get the above error and when I try to save the file, all the code in the file gets deleted. The output logs are very long - here is just a little bit of it:

Traceback (most recent call last):
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\libs\pygls\protocol.py", line 84, in decorator
    self._execute_notification(user_func, *args, **kwargs)
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\libs\pygls\protocol.py", line 228, in _execute_notification
    handler(*params)
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\tool\server.py", line 118, in did_change
    diagnostics: list[Diagnostic] = _linting_helper(document)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\tool\server.py", line 134, in _linting_helper
    return _parse_output_using_regex(result.stdout) if result.stdout else []
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\tool\server.py", line 172, in _parse_output_using_regex
    start = Position(
            ^^^^^^^^^
  File "<attrs generated init lsprotocol.types.Position>", line 6, in __init__
    __attr_validator_character(self, __attr_character, self.character)
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\libs\lsprotocol\validators.py", line 45, in uinteger_validator
    raise ValueError(
ValueError: Position.character should be in range [0:2147483647], but was -1.
Failed to handle notification "textDocument/didSave": DidSaveTextDocumentParams(text_document=TextDocumentIdentifier(uri='file:///n%3A/Florian/Python/UTEC_Tools/modules/classes_data.py'), text=None)
Traceback (most recent call last):
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\libs\pygls\protocol.py", line 316, in _handle_notification
    self._execute_notification(handler, params)
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\libs\pygls\protocol.py", line 228, in _execute_notification
    handler(*params)
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\tool\server.py", line 110, in did_save
    diagnostics: list[Diagnostic] = _linting_helper(document)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\tool\server.py", line 134, in _linting_helper
    return _parse_output_using_regex(result.stdout) if result.stdout else []
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\tool\server.py", line 172, in _parse_output_using_regex
    start = Position(
            ^^^^^^^^^
  File "<attrs generated init lsprotocol.types.Position>", line 6, in __init__
    __attr_validator_character(self, __attr_character, self.character)
  File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2022.0.22-win32-x64\bundled\libs\lsprotocol\validators.py", line 45, in uinteger_validator
    raise ValueError(
ValueError: Position.character should be in range [0:2147483647], but was -1.

**Comment by charliermarsh** [comment]
Do you mind trying with the latest version? Will re-open if you continue to see this!

**Comment by lasinludwig** [comment]
Thanks for getting back to me about this. Just tried out version 2022.2.0 - same issue unfortunately.

**Comment by charliermarsh** [comment]
Oh I'm wondering if this is: https://github.com/charliermarsh/ruff-lsp/pull/51

**Comment by charliermarsh** [comment]
Do you mind trying out the latest version?

**Comment by lasinludwig** [comment]
Still the same issue... maybe it's because i use some other extensions for linting simultaneously... I'll do some tests later and get back to you.

**Comment by jdimmerman** [comment]
@charliermarsh if helpful, I am also seeing this. If I restart the ruff server, I get
```
qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
```

I'm running docker on an M1 mac using an ARM native ubuntu image, which I imagine is the problem. @lasinludwig does that apply to you as well?

**Comment by lasinludwig** [comment]
I'm running it on a Windows machine without any docker stuff... Maybe it's something in my virtual environment?

I tried disabling other linters I run - no effect.

**Comment by charliermarsh** [comment]
I can try getting a Windows VM going on my personal machine. (I don't have a Windows machine to test on, unfortunately.)

@FrancescElies - have you been able to run the Ruff extension on Windows? Just want to check if this is an "all Windows" problem or something more specific.

**Comment by FrancescElies** [comment]
@charliermarsh it works on windows but keep in mind I am not a heavy user of vscode I just happen to have it installed because many of my colleagues use it for coding in python.

Example:
![image](https://user-images.githubusercontent.com/6558089/210595455-605ab0c3-f18d-4a4d-900d-bf185bc88e2e.png)

![image](https://user-images.githubusercontent.com/6558089/210595534-198f01c4-3c9d-45db-98a4-4266439231f0.png)

**Comment by charliermarsh** [comment]
@FrancescElies - Thanks, that's perfect -- I just wanted a quick sanity check to make sure it wasn't entirely busted.

**Comment by niccolomineo** [comment]
> @charliermarsh if helpful, I am also seeing this. If I restart the ruff server, I get
> 
> ```
> qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
> ```
> 
> I'm running docker on an M1 mac using an ARM native ubuntu image, which I imagine is the problem. @lasinludwig does that apply to you as well?

I have your exact issue with the latest `ruff-vscode` version.

**Comment by jdimmerman** [comment]
I worked around the issue by downloading the ruff binary and pointing to it via the ruff path setting

**Comment by niccolomineo** [comment]
Thanks, that's too cumbersome for me as I am working in a docker container via the Dev Containers extension.

 I am resorting to the flake8 / isort extensions, hoping this bug will be eventually fixed properly.

**Comment by jdimmerman** [comment]
I gotchya. Same regarding devcontainer. We just added the binary to the container definition for what it’s worth. A quick fix.

**Comment by charliermarsh** [comment]
Sadly I haven't had a chance to dig into this. @messense - just tagging because this relates to packaging -- anything stand out to you about why folks are running into this?

**Comment by messense** [comment]
> @charliermarsh if helpful, I am also seeing this. If I restart the ruff server, I get
> 
> ```
> qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
> ```
> 
> I'm running docker on an M1 mac using an ARM native ubuntu image, which I imagine is the problem. @lasinludwig does that apply to you as well?

Looks to me that it somehow installed the x86_64 version of Ruff in a arm64 docker container.

**Comment by messense** [comment]
https://github.com/charliermarsh/ruff-vscode/actions/runs/3954289944/jobs/6771477804#step:4:26

So the problem is that you can't build the aarch64 linux extension on a x86_64 linux.

**Comment by charliermarsh** [comment]
Oh interesting. Can that be solved by changing the host machine in the Action?

**Comment by messense** [comment]
We can fix it by installing dependencies via QEMU: https://github.com/charliermarsh/ruff-vscode/pull/119

**Comment by charliermarsh** [comment]
I just cut a new pre-release version with this fix. Do you mind trying it out in your setup?

**Comment by niccolomineo** [comment]
> I just cut a new pre-release version with this fix. Do you mind trying it out in your setup?

I have installed it, but I seem to still have that error:

```
qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
```

**Comment by niccolomineo** [comment]
> I just cut a new pre-release version with this fix. Do you mind trying it out in your setup?

I am not noticing the issue anymore on my machine.

**Comment by charliermarsh** [comment]
Oh, stellar! Thanks @messense.

**Comment by lasinludwig** [comment]
...still the same problem for me... does this output help?

> Traceback (most recent call last):
>   File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2023.5.10291638-win32-x64\bundled\libs\pygls\protocol.py", line 316, in _handle_notification
>     self._execute_notification(handler, params)
>   File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2023.5.10291638-win32-x64\bundled\libs\pygls\protocol.py", line 228, in _execute_notification
>     handler(*params)
>   File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2023.5.10291638-win32-x64\bundled\libs\ruff_lsp\server.py", line 93, in did_save
>     diagnostics: list[Diagnostic] = _linting_helper(document)
>                                     ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2023.5.10291638-win32-x64\bundled\libs\ruff_lsp\server.py", line 117, in _linting_helper
>     return _parse_output_using_regex(result.stdout) if result.stdout else []
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2023.5.10291638-win32-x64\bundled\libs\ruff_lsp\server.py", line 155, in _parse_output_using_regex
>     start = Position(
>             ^^^^^^^^^
>   File "<attrs generated init lsprotocol.types.Position>", line 7, in __init__
>     __attr_validator_character(self, __attr_character, self.character)
>   File "c:\Users\fl\.vscode\extensions\charliermarsh.ruff-2023.5.10291638-win32-x64\bundled\libs\lsprotocol\validators.py", line 45, in uinteger_validator
>     raise ValueError(
> ValueError: Position.character should be in range [0:2147483647], but was -1.
